src/systems/connectors.jl

```diff
@@ -0,0 +1,389 @@
+function with_connector_type(expr)
+    @assert expr isa Expr && (expr.head == :function || (expr.head == :(=) &&
+                                       expr.args[1] isa Expr &&
+                                       expr.args[1].head == :call))
+
+    sig = expr.args[1]
+    body = expr.args[2]
+
+    fname = sig.args[1]
+    args = sig.args[2:end]
+
+    quote
+        function $fname($(args...))
+            function f()
+                $body
+            end
+            res = f()
+            $isdefined(res, :connector_type) && $getfield(res, :connector_type) === nothing ? $Setfield.@set!(res.connector_type = $connector_type(res)) : res
+        end
+    end
+end
+
+macro connector(expr)
+    esc(with_connector_type(expr))
+end
+
+abstract type AbstractConnectorType end
+struct StreamConnector <: AbstractConnectorType end
+struct RegularConnector <: AbstractConnectorType end
+
+function connector_type(sys::AbstractSystem)
+    sts = states(sys)
+    #TODO: check the criteria for stream connectors
+    n_stream = 0
+    n_flow = 0
+    for s in sts
+        vtype = getmetadata(s, ModelingToolkit.VariableConnectType, nothing)
+        vtype === Stream && (n_stream += 1)
+        vtype === Flow   && (n_flow += 1)
+    end
+    (n_stream > 1 && n_flow > 1) && error("There are multiple flow variables in $(nameof(sys))!")
+    n_stream > 1 ? StreamConnector() : RegularConnector()
+end
+
+Base.@kwdef struct Connection
+    inners = nothing
+    outers = nothing
+end
+
+# everything is inner by default until we expand the connections
+Connection(syss) = Connection(inners=syss)
+get_systems(c::Connection) = c.inners
+function Base.in(e::Symbol, c::Connection)
+    f = isequal(e)
+    any(f, c.inners) || any(f, c.outers)
+end
+
+const EMPTY_VEC = []
+
+function Base.show(io::IO, c::Connection)
+    @unpack outers, inners = c
+    if outers === nothing && inners === nothing
+        print(io, "<Connection>")
+    else
+        syss = Iterators.flatten((something(inners, EMPTY_VEC), something(outers, EMPTY_VEC)))
+        splitting_idx = length(inners)
+        sys_str = join((string(nameof(s)) * (i <= splitting_idx ? ("::inner") : ("::outer")) for (i, s) in enumerate(syss)), ", ")
+        print(io, "<", sys_str, ">")
+    end
+end
+
+# symbolic `connect`
+function connect(syss::AbstractSystem...)
+    length(syss) >= 2 || error("connect takes at least two systems!")
+    length(unique(nameof, syss)) == length(syss) || error("connect takes distinct systems!")
+    Equation(Connection(), Connection(syss)) # the RHS are connected systems
+end
```

Comment:
```suggestion
function connect(sys1::AbstractSystem, sys2::AbstractSystem, syss::AbstractSystem...)
    syss = (sys1, sys2, syss...)
    length(unique(nameof, syss)) == length(syss) || error("connect takes distinct systems!")
    Equation(Connection(), Connection(syss)) # the RHS are connected systems
end
```
this avoids introducing methods that takes anything less than two systems